inc/napa/async-sender.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <mutex>
+#include <queue>
+#include <atomic>
+#include <uv.h>
+
+namespace napa{
+
+class AsyncTaskSender {
+public:
+    typedef std::function<void(void)> AsyncTask;
+
+    static AsyncTaskSender* Create(uv_loop_t* loop) {
+        return new AsyncTaskSender(loop);
+    }
+
+    void stop() {
+        needStop_ = true;
+        uv_async_send(&handle_);
+    }
+
+    void post(AsyncTask task) {
+        if (!needStop_) {
+            std::lock_guard<std::mutex> lock(mutex_);
+            queue_.emplace(std::move(task));
+            uv_async_send(&handle_);
+        }
+    }
+
+    static void RunTasks(uv_async_t* handle) {
+        auto sender = reinterpret_cast<AsyncTaskSender*>(handle->data);
+        bool stopped = sender->needStop_;
+        std::queue<AsyncTask> q;
+        {
+            std::lock_guard<std::mutex> lock(sender->mutex_);
+            q.swap(sender->queue_);
+        }
+        while (!q.empty()) {
+            auto task = q.front();
+            q.pop();
+            task();
+        }
+        if (stopped) {
+            uv_close(reinterpret_cast<uv_handle_t*>(&sender->handle_),
+                     [](uv_handle_t* handle) {
+                        auto sender = reinterpret_cast<AsyncTaskSender*>(handle->data);
+                        delete sender;
+                     });
+        }
+    }
+
+    uv_async_t* GetHandle() { return &handle_; }
+    uv_loop_t* GetLoop() { return loop_; }
+
+protected:
+    uv_async_t handle_;
+    uv_loop_t* loop_;
+    std::mutex mutex_;
+    std::queue<AsyncTask> queue_;
+    std::atomic<bool> needStop_;
+
+private:
+    explicit AsyncTaskSender(uv_loop_t* loop)
+            :loop_(loop), needStop_(false) {
+        uv_async_init(loop_, &handle_, RunTasks);
+        handle_.data = static_cast<void*>(this);
+        uv_unref(reinterpret_cast<uv_handle_t*>(&handle_));
+
+        uv_async_send(&handle_);
+    }
+
+    AsyncTaskSender() = delete;
```

Comment:
The default constructor won't be generated give we have the previous constructor implemented, right?